File: frontend/src/types/document.ts
export enum DocumentStatus {
  UPLOADED = 'uploaded',
  PROCESSING = 'processing',
  COMPLETED = 'completed',
  FAILED = 'failed'
}

export enum DocumentType {
  UNKNOWN = 'unknown',
  INVOICE = 'invoice',
  CONTRACT = 'contract',
  RESUME = 'resume',
  RECEIPT = 'receipt',
  FORM = 'form',
  LETTER = 'letter',
  REPORT = 'report',
  OTHER = 'other'
}

export interface Document {
  id: number
  filename: string
  original_filename: string
  file_size: number
  file_type: string
  storage_path: string
  status: DocumentStatus
  extracted_text?: string | null
  page_count?: number | null
  extraction_method?: string | null
  extraction_error?: string | null
  task_id?: string | null
  retry_count: number
  processing_time?: number | null
  document_type: DocumentType
  document_type_confidence?: number | null
  extracted_data?: Record<string, any> | null
  summary?: string | null
  ai_processing_cost?: number | null
  created_at: string
  updated_at: string
}

export interface DocumentListResponse {
  total: number
  items: Document[]
  page: number
  page_size: number
}

export interface SearchResult {
  id: number
  filename: string
  original_filename: string
  document_type: string
  status: string
  summary?: string | null
  created_at: string
  score?: number | null
}

export interface SearchResponse {
  query: string
  total: number
  results: SearchResult[]
  search_type: string
  page: number
  page_size: number
}

export interface WebSocketMessage {
  type: string
  document_id?: number
  status?: string
  message?: string
  progress?: number
  metadata?: Record<string, any>
  timestamp?: number
  connection_id?: string
}

File: backend/app/models/document.py
from sqlalchemy import Column, Integer, String, DateTime, Enum as SQLEnum, BigInteger, Text, Float, JSON, Index, ForeignKey
from sqlalchemy.sql import func
from sqlalchemy.dialects.postgresql import TSVECTOR
from sqlalchemy.orm import relationship
from pgvector.sqlalchemy import Vector
from app.database import Base
import enum


class DocumentStatus(str, enum.Enum):
    UPLOADED = "uploaded"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"


class DocumentType(str, enum.Enum):
    UNKNOWN = "unknown"
    INVOICE = "invoice"
    CONTRACT = "contract"
    RESUME = "resume"
    RECEIPT = "receipt"
    FORM = "form"
    LETTER = "letter"
    REPORT = "report"
    OTHER = "other"


class Document(Base):
    __tablename__ = "documents"
    
    id = Column(Integer, primary_key=True, index=True)
    filename = Column(String(255), nullable=False)
    original_filename = Column(String(255), nullable=False)
    file_size = Column(BigInteger, nullable=False)
    file_type = Column(String(100), nullable=False)
    storage_path = Column(String(500), nullable=False)
    
    # NEW - User ownership
    user_id = Column(Integer, ForeignKey('users.id'), nullable=True, index=True)
    
    status = Column(
        SQLEnum(DocumentStatus),
        default=DocumentStatus.UPLOADED,
        nullable=False,
        index=True
    )
    
    # Extraction fields
    extracted_text = Column(Text, nullable=True)
    page_count = Column(Integer, nullable=True)
    extraction_method = Column(String(50), nullable=True)
    extraction_error = Column(Text, nullable=True)
    
    # Task tracking
    task_id = Column(String(255), nullable=True, index=True)
    retry_count = Column(Integer, default=0)
    processing_time = Column(Float, nullable=True)
    
    # AI fields
    document_type = Column(
        SQLEnum(DocumentType),
        default=DocumentType.UNKNOWN,
        nullable=False,
        index=True
    )
    document_type_confidence = Column(Float, nullable=True)
    extracted_data = Column(JSON, nullable=True)
    summary = Column(Text, nullable=True)
    ai_processing_cost = Column(Float, nullable=True)
    
    # Search fields
    search_vector = Column(TSVECTOR, nullable=True)
    embedding = Column(Vector(384), nullable=True)
    
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)
    
    # Indexes
    __table_args__ = (
        Index('ix_documents_search_vector', 'search_vector', postgresql_using='gin'),
        Index('ix_documents_embedding', 'embedding', postgresql_using='ivfflat', postgresql_ops={'embedding': 'vector_cosine_ops'}),
    )
    
    def __repr__(self):
        return f"<Document(id={self.id}, filename={self.filename}, type={self.document_type}, status={self.status})>"